### PR TITLE
internal/legacy: implement /charm endpoint

### PR DESCRIPTION
**Pull request by rogpeppe**
We use some of the tests from the old server.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/212/

File: internal/legacy/api.go [hunk 1]
```diff
@@ -43,6 +47,17 @@ func (h *Handler) serveCharmInfo(w http.ResponseWriter, req *http.Request) {
...
+	}
+	url, err := charm.ParseReference(strings.TrimPrefix(req.URL.Path, "/"))
+	if err != nil {
+		return errgo.WithCausef(err, params.ErrNotFound, "")
+	}
+	if err := v4.ResolveURL(h.store, url); err != nil {
+		// Note: preserve error cause from resolveURL.
+		return errgo.Mask(err, errgo.Any)
+	}
+	return h.v4.Id["archive"](url, w, req)
```
Comment (urosj) [inline internal/legacy/api.go:62]: Had to look how this h.v4.Id works again. nice mapping :)

File: internal/legacy/api_test.go [hunk 2]
```diff
@@ -52,3 +116,41 @@ func (s *APISuite) TestCharmInfo(c *gc.C) {
 		},
 	})
 }
+
+var serverStatusTests = []struct {
+	path string
+	code int
+}{
+	{"/charm-info/any", 404},
```
Comment (urosj) [inline internal/legacy/api_test.go:124]: Why do you test /charm-info under /charm API?

**Comment by urosj** [comment]
:+1:  (with a small comment)

File: internal/legacy/api.go [hunk 3]
```diff
@@ -8,29 +8,33 @@ import (
...
 	}
-	h.handle("/charm-info", h.serveCharmInfo)
-	h.handle("/charm/", h.serveCharm)
+	h.handle("/charm-info", http.HandlerFunc(h.serveCharmInfo))
+	h.handle("/charm/", router.HandleErrors(h.serveCharm))
 	return h
 }
 
-func (h *Handler) handle(path string, handler http.HandlerFunc) {
+func (h *Handler) handle(path string, handler http.Handler) {
```
Comment (frankban) [inline internal/legacy/api.go:37]: Nice change,

File: internal/v4/archive.go [hunk 4]
```diff
@@ -53,11 +53,13 @@ func (h *Handler) serveArchive(id *charm.Reference, w http.ResponseWriter, req *
...
 	r, size, err := h.openBlob(id)
 	if err != nil {
-		return errgo.Mask(err)
+		return errgo.Mask(err, errgo.Is(params.ErrNotFound))
 	}
 	defer r.Close()
 	// TODO frankban 2014-08-22: log possible IncCounter errors.
 	go h.store.IncCounter(entityStatsKey(id, params.StatsArchiveDownload))
+	// TODO(rog) should we set connection=close here?
+	// See https://codereview.appspot.com/5958045
```
Comment (frankban) [inline internal/v4/archive.go:62]: Thanks for this comment, interesting.

**Comment by frankban** [comment]
Thanks for this branch Roger, :+1: land as is.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**